File: bot/domain/parsers/command_parser.py
"""Tokenises command text against a CommandConfig regex into a ParsedCommand."""

import re
from typing import TYPE_CHECKING, ClassVar

from bot.domain.parsers.command_parser_regex import CommandParserRegex

if TYPE_CHECKING:
    from bot.domain.commands.base import CommandConfig, OptionDef, ParsedCommand


class CommandParser:
    LEADING_COMMA: ClassVar[re.Pattern[str]] = re.compile(r'^\s*,\s*')

    def __init__(self, config: 'CommandConfig') -> None:
        self._config = config
        self._regex = CommandParserRegex.build(config)

    def matches(self, text: str) -> bool:
        return self._regex.search(text) is not None

    def parse(self, text: str) -> 'ParsedCommand':
        from bot.domain.commands.base import ParsedCommand

        remaining = self.LEADING_COMMA.sub('', text)
        command_name, remaining = self._consume_command_name(remaining)

        flags: set[str] = set()
        options: dict[str, str] = {}
        rest_parts: list[str] = []

        tokens = [t for t in remaining.split() if t]
        index = 0
        while index < len(tokens):
            index = self._consume_token(tokens, index, options, flags, rest_parts)

        return ParsedCommand(
            command_name=command_name,
            flags=flags,
            options=options,
            rest=' '.join(rest_parts),
        )

    def _consume_command_name(self, text: str) -> tuple[str, str]:
        for name in CommandParserRegex.sorted_names(self._config):
            pattern = CommandParserRegex.name_pattern(name)
            match = re.match(pattern, text, re.IGNORECASE)
            if not match:
                continue
            return name, text[match.end() :].strip()
        return '', text

    def _consume_token(
        self,
        tokens: list[str],
        index: int,
        options: dict[str, str],
        flags: set[str],
        rest_parts: list[str],
    ) -> int:
        token = tokens[index]
        opt = self._try_match_option(token, options)
        if opt:
            options[opt[0]] = opt[1]
            return index + 1

        positional = self._try_positional_option(tokens, index, options)
        if positional:
            options[token] = positional
            return index + 2

        flag = self._try_match_flag(token, flags)
        if flag:
            flags.add(flag)
            return index + 1

        rest_parts.append(token)
        return index + 1

    def _try_positional_option(
        self,
        tokens: list[str],
        index: int,
        options: dict[str, str],
    ) -> str | None:
        if index + 1 >= len(tokens):
            return None
        value = self._try_match_option_as_value(tokens[index + 1])
        if value is None or value in options:
            return None
        return value

    def _try_match_option_as_value(self, token: str) -> str | None:
        for opt in self._config.options:
            for value in opt.values:
                if re.match(f'^{CommandParserRegex.escape(value)}$', token, re.IGNORECASE):
                    return value
        return None

    def _try_match_option(self, token: str, matched: dict[str, str]) -> tuple[str, str] | None:
        for opt in self._config.options:
            if opt.name in matched:
                continue
            value = self._match_option_value(opt, token)
            if value is not None:
                return opt.name, value
        return None

    @staticmethod
    def _match_option_value(opt: 'OptionDef', token: str) -> str | None:
        for value in opt.values:
            if re.match(f'^{CommandParserRegex.escape(value)}$', token, re.IGNORECASE):
                return value
        if opt.pattern and re.match(f'^{opt.pattern}$', token, re.IGNORECASE):
            return token
        return None

    def _try_match_flag(self, token: str, matched: set[str]) -> str | None:
        for flag in self._config.flags:
            if flag in matched:
                continue
            if re.match(f'^{CommandParserRegex.escape(flag)}$', token, re.IGNORECASE):
                return flag
        return None

File: gateway/src/parsers/CommandParser.ts
import { ArgType } from '../types/commandConfig.js';
import type { CommandConfig, ParsedCommand } from '../types/commandConfig.js';

export default class CommandParser {
  private readonly regex: RegExp;

  constructor(private readonly config: CommandConfig) {
    this.regex = this.buildRegex();
  }

  matches(text: string): boolean {
    return this.regex.test(text);
  }

  parse(text: string): ParsedCommand {
    const { commandName, remaining } = this.extractCommandName(text.replace(/^\s*,\s*/, ''));

    const tokens = remaining.split(/\s+/).filter((t) => t.length > 0);
    const flags = new Set<string>();
    const options = new Map<string, string>();
    const restParts: string[] = [];

    for (const token of tokens) {
      if (this.tryMatchOption(token, options)) continue;
      if (this.tryMatchFlag(token, flags)) continue;
      restParts.push(token);
    }

    return { commandName, flags, options, rest: restParts.join(' ') };
  }

  private extractCommandName(remaining: string): { commandName: string; remaining: string } {
    const names = [this.config.name, ...(this.config.aliases || [])];
    for (const name of names) {
      const namePattern = this.replaceDiacritics(name).replaceAll(/\s+/g, String.raw`\s*`);
      const match = new RegExp(`^${namePattern}`, 'i').exec(remaining);
      if (match) {
        return { commandName: name, remaining: remaining.slice(match[0].length).trim() };
      }
    }
    return { commandName: '', remaining };
  }

  private tryMatchOption(token: string, options: Map<string, string>): boolean {
    for (const opt of this.config.options || []) {
      if (options.has(opt.name)) continue;
      if (opt.values) {
        const matchedValue = opt.values.find((v) =>
          new RegExp(`^${this.replaceDiacritics(v)}$`, 'i').test(token),
        );
        if (matchedValue) {
          options.set(opt.name, matchedValue);
          return true;
        }
      }
      if (opt.pattern && new RegExp(`^${opt.pattern}$`, 'i').test(token)) {
        options.set(opt.name, token);
        return true;
      }
    }
    return false;
  }

  private tryMatchFlag(token: string, flags: Set<string>): boolean {
    const matchedFlag = (this.config.flags || []).find((f) => {
      if (flags.has(f)) return false;
      return new RegExp(`^${this.replaceDiacritics(f)}$`, 'i').test(token);
    });
    if (matchedFlag) {
      flags.add(matchedFlag);
      return true;
    }
    return false;
  }

  private buildRegex(): RegExp {
    const parts: string[] = [];

    parts.push(String.raw`^\s*,\s*`);

    const names = [this.config.name, ...(this.config.aliases || [])];
    const namePatterns = names.map((n) =>
      this.replaceDiacritics(n).replaceAll(/\s+/g, String.raw`\s*`),
    );
    if (namePatterns.length === 1) {
      parts.push(namePatterns[0]);
    } else {
      parts.push(`(?:${namePatterns.join('|')})`);
    }

    for (const opt of this.config.options || []) {
      if (opt.values) {
        const sorted = [...opt.values].sort((a, b) => b.length - a.length);
        const valPatterns = sorted.map((v) => this.replaceDiacritics(v));
        parts.push(String.raw`\s*(?:${valPatterns.join('|')})?`);
      } else if (opt.pattern) {
        parts.push(String.raw`\s*(?:${opt.pattern})?`);
      }
    }

    if (this.config.flags && this.config.flags.length > 0) {
      const flagPatterns = this.config.flags.map((f) => this.replaceDiacritics(f));
      parts.push(String.raw`(?:\s+(?:${flagPatterns.join('|')}))*`);
    }

    const args = this.config.args ?? ArgType.None;
    switch (args) {
      case ArgType.None:
        parts.push(String.raw`\s*$`);
        break;
      case ArgType.Required:
        if (this.config.argsPattern) {
          const src = this.config.argsPattern.source.replace(/^\^/, '').replace(/\$$/, '');
          parts.push(String.raw`\s*${src}\s*$`);
        } else {
          parts.push(String.raw`\s+.+`);
        }
        break;
      case ArgType.Optional:
        if (this.config.argsPattern) {
          const src = this.config.argsPattern.source.replace(/^\^/, '').replace(/\$$/, '');
          parts.push(String.raw`\s*${src}\s*$`);
        } else {
          parts.push(String.raw`(?:\s+.*)?$`);
        }
        break;
    }

    return new RegExp(parts.join(''), 'i');
  }

  private replaceDiacritics(s: string): string {
    return s.replaceAll(/[.*+?^${}()|[\]\\]/g, String.raw`\$&`).replaceAll(/[^\x00-\x7F]/g, '.');
  }
}
